### PR TITLE
[GH-745] Check status of the Jira instance before trying to install

### PR DESCRIPTION
**Pull request by sonereker**
Signed-off-by: Soner Eker <sonereker@gmail.com>

#### Summary
Before trying to install Jira instance (cloud or server), do a `GET` request to instance `/status`
 endpoint which basically provides state information of the Jira instance. If instance is in `RUNNING` state, it means instance is accessible and it's OK to continue with installation.

#### Ticket Link
https://github.com/mattermost/mattermost-plugin-jira/issues/745

#### Verification Video:

https://drive.google.com/file/d/1SDyHiRCy5ryKxCnKU98xvokVH5cD677R/view?usp=sharing

### CODE REVIEW

File: server/utils/utils.go [hunk 1]
```diff
@@ -65,3 +68,41 @@ func IsJiraCloudURL(jiraURL string) (bool, error) {
...
+	if err != nil {
+		return false, nil
+	}
+	var j JiraStatus
+	err = json.Unmarshal(resBody, &j)
+	if err != nil {
+		return false, nil
+	}
+
+	return j.State == "RUNNING", nil
```
Comment (mickmister) [inline server/utils/utils.go:107]: The video the PR description is tested with Jira Cloud. Has this PR also been tested with a Jira Server instance? You can spin up a Jira Server instance on port `8080` by running `docker-compose up` in the root directory of this repo.

File: server/command.go [hunk 2]
```diff
@@ -852,6 +861,17 @@ func executeInstanceInstallServer(p *Plugin, c *plugin.Context, header *model.Co
 	})
 }
 
+func checkIfJiraIsAccessible(jiraURL string) (string, bool) {
+	jiraIsAccessible, err := utils.IsJiraAccessible(jiraURL)
+	if err != nil {
+		return err.Error(), false
+	}
+	if !jiraIsAccessible {
+		return "It looks like we couldn't validate the connection to your Jira server. Please make sure the URL was entered correctly. This could also be because of existing firewall or proxy rules. If you intend to have a one way integration from Jira to Mattermost this is not an issue.", false
```
Comment (mickmister) [inline server/command.go:870]: Can we have this constant at the top of the file?

File: server/utils/utils.go [hunk 3]
```diff
@@ -65,3 +68,41 @@ func IsJiraCloudURL(jiraURL string) (bool, error) {
...
+
+// IsJiraAccessible checks if `/status` endpoint of the Jira URL is accessible
+// and responding with the correct state which is "RUNNING"
+func IsJiraAccessible(jiraURL string) (bool, error) {
+	u, err := url.Parse(jiraURL)
+	if err != nil {
+		return false, nil
+	}
+	if u.Host == "" {
+		return false, errors.Errorf("Invalid URL, no hostname: %q", jiraURL)
```
Comment (mickmister) [inline server/utils/utils.go:84]: Seems odd that this function has an error return value, but is only returning an error in one case. For instance for the `r.StatusCode != http.StatusOK` block, we could return to the admin that their `Jira server returned http status code 404 when checking for availability`.

**Review by mickmister — CHANGES_REQUESTED** [review]
Awesome work @sonereker! Thanks for this contribution.

I just have a few requests and one question regarding Jira Server.

File: server/utils/utils.go (hunk 1 above)
Comment (sonereker) [reply server/utils/utils.go]: Updated video.

File: server/command.go (hunk 2 above)
Comment (sonereker) [reply server/command.go]: Done.

File: server/utils/utils.go (hunk 3 above)
Comment (sonereker) [reply server/utils/utils.go:84]: added that one, and also added below one;

`Jira server is not in correct state, it should be up and running`

**Comment by sonereker** [comment]
@mickmister thanks for reviewing, fixed all, added some more tests.

**Review by mickmister — APPROVED** [review]
LGTM :+1:

Thanks @sonereker!

**Review by jfrerich — APPROVED** [review]
code and response messages LGTM!  Thanks!

**Comment by DHaussermann** [comment]
@mickmister I'd like to get this merged. Can you please see if you're able to pull `master` up into this branch?

**Comment by sonereker** [comment]
@DHaussermann @mickmister working on it.

**Comment by sonereker** [comment]
@DHaussermann fixed the conflict.

**Review by mickmister — APPROVED** [review]

**Review by DHaussermann — APPROVED** [review]
Tested and passed
This is working as expected in all cases tested:
- URL is to a vlid instance
- URL 404s
- URL is reachable but is not a Jira instance
- URL format is invalid
- Install a working Jira cloud instance
- Install a working Jira server instance

LGTM!

Huge thanks @sonereker for this enhancement! Sorry for the delay in getting this approved.

**Comment by mickmister** [comment]
@sonereker Sorry for missing this approved PR! Thanks for this great contribution 🎉

**Pull request merged**